## Superstore Exploratory Data Analysis

## Data Dictionary

| Column Name   | Description |
|---|---|
| **Row ID** | Unique identifier for each row of data. |
| **Order ID** | Unique order identifier for each customer. |
| **Order Date** | Date when the product was ordered. |
| **Ship Date** | Date when the product was shipped. |
| **Ship Mode** | Shipping method selected by the customer (e.g., First Class, Second Class, Standard Class, Same Day). |
| **Customer ID** | Unique identifier for each customer. |
| **Customer Name** | Name of the customer. |
| **Segment** | Customer segment (Consumer, Corporate, Home Office). |
| **Country** | Country of customer residence. |
| **City** | City of customer residence. |
| **State** | State or region of the customer. |
| **Postal Code** | Postal/ZIP code of the customer. |
| **Region** | Geographic region where the customer belongs (e.g., East, West, Central, South). |
| **Product ID** | Unique product identifier. |
| **Category** | Main product category (Furniture, Office Supplies, Technology). |
| **Sub-Category** | Product sub-category (e.g., Chairs, Binders, Phones). |
| **Product Name** | Product name. |
| **Sales** | Sales revenue (amount received). |
| **Quantity** | Quantity of products sold. |
| **Discount** | Discount applied to the product (0–1). |
| **Profit** | Profit (or loss) generated from the order line. |

📎 **Sample Dataset File**: [Superstore Sample Dataset (Excel)](https://community.tableau.com/s/question/0D54T00000CWeX8SAL/sample-superstore-sales-excelxls)

# 1. Import Required Libraries

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import plotly.express as px
import plotly.graph_objects as go
from plotly.subplots import make_subplots

import warnings
warnings.filterwarnings("ignore", category=UserWarning)
warnings.filterwarnings("ignore", category=FutureWarning)

# 2. Load and Inspect Data

In [ ]:
df = pd.read_excel("/home/ducanhhh/superstore/data/superstore.xlsx", sheet_name=0)
pd.set_option('display.max_columns', None)
df.head()

Row ID        Order ID Order Date  Ship Date       Ship Mode Customer ID  \
0       1  CA-2016-152156 2016-11-08 2016-11-11    Second Class    CG-12520   
1       2  CA-2016-152156 2016-11-08 2016-11-11    Second Class    CG-12520   
2       3  CA-2016-138688 2016-06-12 2016-06-16    Second Class    DV-13045   
3       4  US-2015-108966 2015-10-11 2015-10-18  Standard Class    SO-20335   
4       5  US-2015-108966 2015-10-11 2015-10-18  Standard Class    SO-20335   

     Customer Name    Segment        Country             City       State  \
0      Claire Gute   Consumer  United States        Henderson    Kentucky   
1      Claire Gute   Consumer  United States        Henderson    Kentucky   
2  Darrin Van Huff  Corporate  United States      Los Angeles  California   
3   Sean O'Donnell   Consumer  United States  Fort Lauderdale     Florida   
4   Sean O'Donnell   Consumer  United States  Fort Lauderdale     Florida   

   Postal Code Region       Product ID         Category Sub-Category  \
0        42420  South  FUR-BO-10001798        Furniture    Bookcases   
1        42420  South  FUR-CH-10000454        Furniture       Chairs   
2        90036   West  OFF-LA-10000240  Office Supplies       Labels   
3        33311  South  FUR-TA-10000577        Furniture       Tables   
4        33311  South  OFF-ST-10000760  Office Supplies      Storage   

                                        Product Name     Sales  Quantity  \
0                  Bush Somerset Collection Bookcase  261.9600         2   
1  Hon Deluxe Fabric Upholstered Stacking Chairs,...  731.9400         3   
2  Self-Adhesive Address Labels for Typewriters b...   14.6200         2   
3      Bretford CR4500 Series Slim Rectangular Table  957.5775         5   
4                     Eldon Fold 'N Roll Cart System   22.3680         2   

   Discount    Profit  
0      0.00   41.9136  
1      0.00  219.5820  
2      0.00    6.8714  
3      0.45 -383.0310  
4      0.20    2.5164

## 2.1 Data Information

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9994 entries, 0 to 9993
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype         
---  ------         --------------  -----         
 0   Row ID         9994 non-null   int64         
 1   Order ID       9994 non-null   object        
 2   Order Date     9994 non-null   datetime64[ns]
 3   Ship Date      9994 non-null   datetime64[ns]
 4   Ship Mode      9994 non-null   object        
 5   Customer ID    9994 non-null   object        
 6   Customer Name  9994 non-null   object        
 7   Segment        9994 non-null   object        
 8   Country        9994 non-null   object        
 9   City           9994 non-null   object        
 10  State          9994 non-null   object        
 11  Postal Code    9994 non-null   int64         
 12  Region         9994 non-null   object        
 13  Product ID     9994 non-null   object        
 14  Category       9994 non-null   object        
 15  Sub-Category   9994 n

In [4]:
df['Order Date'] = pd.to_datetime(df['Order Date'])
df['Ship Date'] = pd.to_datetime(df['Ship Date'])

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9994 entries, 0 to 9993
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype         
---  ------         --------------  -----         
 0   Row ID         9994 non-null   int64         
 1   Order ID       9994 non-null   object        
 2   Order Date     9994 non-null   datetime64[ns]
 3   Ship Date      9994 non-null   datetime64[ns]
 4   Ship Mode      9994 non-null   object        
 5   Customer ID    9994 non-null   object        
 6   Customer Name  9994 non-null   object        
 7   Segment        9994 non-null   object        
 8   Country        9994 non-null   object        
 9   City           9994 non-null   object        
 10  State          9994 non-null   object        
 11  Postal Code    9994 non-null   int64         
 12  Region         9994 non-null   object        
 13  Product ID     9994 non-null   object        
 14  Category       9994 non-null   object        
 15  Sub-Category   9994 n

# 3. Data Cleaning

## 3.1 Check for Duplicates and Null Values

Check duplicates

In [5]:
print(f"Number of duplicates: {df.duplicated().sum()}")
df = df.drop_duplicates()

Number of duplicates: 0


Check nulls

In [6]:
print(f"Number of null values: {df.isnull().sum().sum()}")

Number of null values: 0


## 3.2 Drop Unnecessary Columns

Drop 'Row ID' and 'Country' (assuming single country based on dataset)

In [7]:
df = df.drop(['Row ID', 'Country'], axis=1)
df.head(2)

Order ID Order Date  Ship Date     Ship Mode Customer ID  \
0  CA-2016-152156 2016-11-08 2016-11-11  Second Class    CG-12520   
1  CA-2016-152156 2016-11-08 2016-11-11  Second Class    CG-12520   

  Customer Name   Segment       City     State  Postal Code Region  \
0   Claire Gute  Consumer  Henderson  Kentucky        42420  South   
1   Claire Gute  Consumer  Henderson  Kentucky        42420  South   

        Product ID   Category Sub-Category  \
0  FUR-BO-10001798  Furniture    Bookcases   
1  FUR-CH-10000454  Furniture       Chairs   

                                        Product Name   Sales  Quantity  \
0                  Bush Somerset Collection Bookcase  261.96         2   
1  Hon Deluxe Fabric Upholstered Stacking Chairs,...  731.94         3   

   Discount    Profit  
0       0.0   41.9136  
1       0.0  219.5820

## 3.3 Data Overview

In [8]:
df.describe()

Order Date                      Ship Date  \
count                           9994                           9994   
mean   2016-04-30 00:07:12.259355648  2016-05-03 23:06:58.571142912   
min              2014-01-03 00:00:00            2014-01-07 00:00:00   
25%              2015-05-23 00:00:00            2015-05-27 00:00:00   
50%              2016-06-26 00:00:00            2016-06-29 00:00:00   
75%              2017-05-14 00:00:00            2017-05-18 00:00:00   
max              2017-12-30 00:00:00            2018-01-05 00:00:00   
std                              NaN                            NaN   

        Postal Code         Sales     Quantity     Discount       Profit  
count   9994.000000   9994.000000  9994.000000  9994.000000  9994.000000  
mean   55190.379428    229.858001     3.789574     0.156203    28.656896  
min     1040.000000      0.444000     1.000000     0.000000 -6599.978000  
25%    23223.000000     17.280000     2.000000     0.000000     1.728750  
50%    56430.500000     54.490000     3.000000     0.200000     8.666500  
75%    90008.000000    209.940000     5.000000     0.200000    29.364000  
max    99301.000000  22638.480000    14.000000     0.800000  8399.976000  
std    32063.693350    623.245101     2.225110     0.206452   234.260108

## Key Observations

- **Sales (Revenue)**: Mean ≈ 230, max ≈ 22,638. Indicates outliers in order values.
- **Profit**: Min -6,599, shows severe losses. Priority for investigation (discounts, products, regions).
- **Discount**: 0% to 80%. Analyze relation with Profit for optimal thresholds.

# 4. Exploratory Data Analysis (EDA)

## Key Features & Analysis Directions for Retail

- **Sales & Profit**: Identify top contributors; detect losses.
- **Customer Segmentation**: Prioritize high-value groups.
- **Geographic Sales**: Trends and opportunities by Region/City/State.
- **Time Series**: Seasonality, growth via Order/Ship Dates.
- **Discount Impact**: Relation with Profit to avoid losses.

## Sales and Profit Analysis

### Product Category Assessment

In [9]:
# 1. Quantity (sorted by count)
type_of_products = df['Category'].value_counts().reset_index()
type_of_products.columns = ['Types of Products', 'count']

# 2. Sales (sorted by sales)
products_w_sales = df.groupby('Category')['Sales'].sum().sort_values(ascending=False).reset_index()
products_w_sales.columns = ['Types of Products', 'total_sales']

# 3. Profit (sorted by profit)
products_w_profit = df.groupby('Category')['Profit'].sum().sort_values(ascending=False).reset_index()
products_w_profit.columns = ['Types of Products', 'total_profit']

print("PRODUCT DISTRIBUTION BY QUANTITY".center(50))
display(type_of_products)

print("\nPRODUCT DISTRIBUTION BY REVENUE".center(50))
display(products_w_sales)

print("\nPRODUCT DISTRIBUTION BY PROFIT".center(50))
display(products_w_profit)


         PRODUCT DISTRIBUTION BY QUANTITY         


Types of Products  count
0   Office Supplies   6026
1         Furniture   2121
2        Technology   1847

         
PRODUCT DISTRIBUTION BY REVENUE         


Types of Products  total_sales
0        Technology  836154.0330
1         Furniture  741999.7953
2   Office Supplies  719047.0320

         
PRODUCT DISTRIBUTION BY PROFIT          


Types of Products  total_profit
0        Technology   145454.9481
1   Office Supplies   122490.8008
2         Furniture    18451.2728

In [10]:
theme = 'plotly_white'
color_map = {
    'Technology': '#b1e7cd',
    'Furniture': '#6C8C9E',
    'Office Supplies': '#be9b7b'
}

fig = make_subplots(
    rows=1, cols=3,
    subplot_titles=(
        "PRODUCT DISTRIBUTION BY SALES QUANTITY",
        "PRODUCT DISTRIBUTION BY REVENUE",
        "PRODUCT DISTRIBUTION BY PROFIT"
    ),
    horizontal_spacing=0.12
)

# 1. Quantity
total_count = type_of_products['count'].sum()
fig.add_trace(
    go.Bar(
        x=type_of_products['Types of Products'],
        y=type_of_products['count'],
        text=[f'{c}<br>({c/total_count:.1%})' for c in type_of_products['count']],
        textposition='auto',
        marker_color=[color_map[x] for x in type_of_products['Types of Products']],
        name='Qty'
    ),
    row=1, col=1
)

# 2. Sales
fig.add_trace(
    go.Bar(
        x=products_w_sales['Types of Products'],
        y=products_w_sales['total_sales'],
        text=products_w_sales['total_sales'].map(lambda x: f'${x:,.0f}'),
        textposition='auto',
        marker_color=[color_map[x] for x in products_w_sales['Types of Products']],
        name='Revenue'
    ),
    row=1, col=2
)

# 3. Profit
fig.add_trace(
    go.Bar(
        x=products_w_profit['Types of Products'],
        y=products_w_profit['total_profit'],
        text=products_w_profit['total_profit'].map(lambda x: f'${x:,.0f}'),
        textposition='auto',
        marker_color=[color_map[x] for x in products_w_profit['Types of Products']],
        name='Profit'
    ),
    row=1, col=3
)


fig.update_layout(
    template=theme,
    height=600,
    width=1600,
    showlegend=False,
    font=dict(size=12),
    margin=dict(l=40, r=40, t=80, b=60)
)

fig.update_xaxes(title_text="Product Type", row=1, col=1)
fig.update_yaxes(title_text="Sales Quantity", row=1, col=1)
fig.update_xaxes(title_text="Product Type", row=1, col=2)
fig.update_yaxes(title_text="Revenue ($)", row=1, col=2)
fig.update_xaxes(title_text="Product Type", row=1, col=3)
fig.update_yaxes(title_text="Profit ($)", row=1, col=3)

fig.show()

#### Office Supplies
- Highest volume (60%+ orders), but low margin. Stable cash flow, not growth driver.
 
#### Furniture
- High revenue (~742K), low profit (~18K, ~2.5% margin). Risks: costs, discounts. Needs optimization.
 
#### Technology
- 18% orders, highest revenue/profit. High margin, main profit contributor. Invest more.

### Detailed Analysis of Sub-categories

Aggregate by Category and Sub-Category

In [11]:
subcat_summary = df.groupby(['Category', 'Sub-Category']).agg(
    total_sales=('Sales', 'sum'),
    total_profit=('Profit', 'sum')
).sort_values(by='total_sales', ascending=False).reset_index()

Sales Bar

In [12]:
fig_sales_subcat = px.bar(
    subcat_summary.sort_values(by='total_sales', ascending=True),
    x='total_sales',
    y='Sub-Category',
    color='Category',
    title='TOTAL REVENUE BY SUB-CATEGORY AND MAIN CATEGORY',
    orientation='h',
    color_discrete_map=color_map,
    template=theme
)
fig_sales_subcat.update_layout(xaxis_title='Total Revenue', yaxis_title='Sub-Category')
fig_sales_subcat.show()

Profit Bar

In [13]:
fig_profit_subcat = px.bar(
    subcat_summary.sort_values(by='total_profit', ascending=True),
    x='total_profit',
    y='Sub-Category',
    color='Category',
    title='TOTAL PROFIT BY SUB-CATEGORY AND MAIN CATEGORY',
    orientation='h',
    color_discrete_map=color_map,
    template=theme
)
fig_profit_subcat.update_layout(xaxis_title='Total Profit', yaxis_title='Sub-Category')
fig_profit_subcat.show()

### Overview
- **Phones**: Top sales (330K), profit (44.5K).
- **Chairs**: 328K sales, Furniture leader.
- **Storage/Binders**: Stable revenue in Office Supplies.
- **Tables**: 207K sales, but -17.7K profit (discounts/costs?).

Profit Boxplot by Sub-Category

In [14]:
fig_profit_box = px.box(
    df,
    x='Sub-Category',
    y='Profit',
    color='Category',
    title='PROFIT DISTRIBUTION BY SUB-CATEGORY',
    template=theme,
    color_discrete_map=color_map
)
fig_profit_box.update_layout(xaxis_title='Sub-Category', yaxis_title='Profit')
fig_profit_box.show()

### Conclusion
- **Technology**: Leading profit; Copiers high/stable, Machines variable.
- **Office Supplies**: Low avg profit, high volatility (Binders/Supplies).
- **Furniture**: Weakest; Tables/Bookcases often loss-making. Optimize costs/products.

## Trend Analysis

Monthly trends

In [15]:
time_trend = df.groupby(pd.Grouper(key='Order Date', freq='M')).agg(
    total_sales=('Sales', 'sum'),
    total_profit=('Profit', 'sum')
).reset_index()

fig_time = px.line(
    time_trend,
    x='Order Date',
    y=['total_sales', 'total_profit'],
    title='Revenue and Profit Trends Over Time (Monthly)',
    template=theme
)
fig_time.update_layout(xaxis_title='Date', yaxis_title='Amount', legend_title='Metrics')
fig_time.show()

### Trends
- Fluctuations; spikes (e.g., Mar 2014: 55K sales, low profit).
- Imbalance: High sales not always high profit (costs/discounts).
- Focus: Cost management, pricing for stable margins.

## Geographic Analysis of Sales and Profit

In [16]:
color_map_region = {
    'Central': '#b1e7cd',
    'East': '#fff4e6',
    'South': '#be9b7b',
    'West': '#add8e6'
}


Aggregate by Region

In [17]:
region_summary = df.groupby('Region').agg(
    total_sales=('Sales', 'sum'),
    total_profit=('Profit', 'sum')
).reset_index()

Sorted for plots

In [18]:
sorted_sales = region_summary.sort_values(by='total_sales', ascending=True)
sorted_profit = region_summary.sort_values(by='total_profit', ascending=True)

Subplots

In [19]:
fig_region = make_subplots(
    rows=1, cols=2,
    subplot_titles=("Total Revenue by Region", "Total Profit by Region"),
    horizontal_spacing=0.15
)

Profit & Sales Bar

In [20]:
fig_region.add_trace(
    go.Bar(
        x=sorted_sales['total_sales'],
        y=sorted_sales['Region'],
        orientation='h',
        marker_color=[color_map_region[r] for r in sorted_sales['Region']],
        name='Total Sales',
        text=sorted_sales['total_sales'].round(2),
        textposition='auto'
    ),
    row=1, col=1
)

fig_region.add_trace(
    go.Bar(
        x=sorted_profit['total_profit'],
        y=sorted_profit['Region'],
        orientation='h',
        marker_color=[color_map_region[r] for r in sorted_profit['Region']],
        name='Total Profit',
        text=sorted_profit['total_profit'].round(2),
        textposition='auto'
    ),
    row=1, col=2
)

fig_region.update_layout(
    template=theme,
    height=600,
    showlegend=False,
    font=dict(size=12),
    margin=dict(l=40, r=40, t=80, b=60)
)
fig_region.update_xaxes(title_text="Total Revenue ($)", row=1, col=1)
fig_region.update_yaxes(title_text="Region", row=1, col=1)
fig_region.update_xaxes(title_text="Total Profit ($)", row=1, col=2)
fig_region.update_yaxes(title_text="Region", row=1, col=2)
fig_region.show()

### Analysis
- **West**: Top sales (725K), profit (108K).
- **East**: Strong (679K sales, 92K profit).
- **Central**: 501K sales, 40K profit (lower efficiency).
- **South**: 392K sales, 47K profit (efficient for scale).

State abbreviations map

In [21]:
us_state_to_abbrev = {
    "Alabama": "AL", "Alaska": "AK", "Arizona": "AZ", "Arkansas": "AR", "California": "CA",
    "Colorado": "CO", "Connecticut": "CT", "Delaware": "DE", "Florida": "FL", "Georgia": "GA",
    "Hawaii": "HI", "Idaho": "ID", "Illinois": "IL", "Indiana": "IN", "Iowa": "IA",
    "Kansas": "KS", "Kentucky": "KY", "Louisiana": "LA", "Maine": "ME", "Maryland": "MD",
    "Massachusetts": "MA", "Michigan": "MI", "Minnesota": "MN", "Mississippi": "MS", "Missouri": "MO",
    "Montana": "MT", "Nebraska": "NE", "Nevada": "NV", "New Hampshire": "NH", "New Jersey": "NJ",
    "New Mexico": "NM", "New York": "NY", "North Carolina": "NC", "North Dakota": "ND", "Ohio": "OH",
    "Oklahoma": "OK", "Oregon": "OR", "Pennsylvania": "PA", "Rhode Island": "RI", "South Carolina": "SC",
    "South Dakota": "SD", "Tennessee": "TN", "Texas": "TX", "Utah": "UT", "Vermont": "VT",
    "Virginia": "VA", "Washington": "WA", "West Virginia": "WV", "Wisconsin": "WI", "Wyoming": "WY"
}

df['State_abb'] = df['State'].map(us_state_to_abbrev)

Aggregate by State

In [22]:
state_summary = df.groupby('State_abb').agg(
    total_sales=('Sales', 'sum'),
    total_profit=('Profit', 'sum')
).reset_index()

Choropleth subplots

In [23]:
fig_state = make_subplots(
    rows=1, cols=2,
    subplot_titles=("Total Revenue by State", "Total Profit by State"),
    specs=[[{'type': 'choropleth'}, {'type': 'choropleth'}]],
    horizontal_spacing=0.25
)

Sales & Profit Choropleth

In [24]:
fig_state.add_trace(
    go.Choropleth(
        locations=state_summary['State_abb'],
        z=state_summary['total_sales'],
        locationmode='USA-states',
        colorscale='teal',
        colorbar_title='Revenue ($)',
        colorbar_x=0.45
    ),
    row=1, col=1
)

fig_state.add_trace(
    go.Choropleth(
        locations=state_summary['State_abb'],
        z=state_summary['total_profit'],
        locationmode='USA-states',
        colorscale='tropic_r',
        zmin=-30000, zmax=30000,
        colorbar_title='Profit ($)',
        colorbar_x=1.02
    ),
    row=1, col=2
)

fig_state.update_layout(
    title_text="REVENUE AND PROFIT ANALYSIS BY STATE",
    title_x=0.5,
    geo1=dict(scope='usa', showlakes=True, lakecolor='rgb(105,192,249)'),
    geo2=dict(scope='usa', showlakes=True, lakecolor='rgb(105,192,249)'),
    font=dict(size=14),
    template=theme,
    height=600,
    margin=dict(l=30, r=30, t=80, b=20)
)
fig_state.show()

### Revenue by State
- High: CA, NY, TX, PA (dense population).
- Low: Central/Northwest.
 
### Profit by State
- Positive: CA, NY, MI.
- Negative: TX, CO, AZ, TN, MS (costs/discounts).
- High revenue, low profit: Efficiency issues.

## Customer Segment Analysis - Sales and Profit

Aggregate by Segment

In [25]:
segment_summary = df.groupby('Segment').agg(
    total_sales=('Sales', 'sum'),
    total_profit=('Profit', 'sum')
).sort_values(by='total_sales', ascending=False)

Cumulative percentages

In [26]:
segment_summary['cumperc_sales'] = segment_summary['total_sales'].cumsum() / segment_summary['total_sales'].sum() * 100
segment_summary['cumperc_profit'] = segment_summary['total_profit'].cumsum() / segment_summary['total_profit'].sum() * 100

Plot

In [27]:
fig_seg = make_subplots(specs=[[{"secondary_y": True}]])

colors = ['#5dade2', '#58d68d']

Bars & Lines

Lines

In [28]:
fig_seg.add_trace(go.Bar(x=segment_summary.index, y=segment_summary['total_sales'], name="Sales", marker_color=colors[0]), secondary_y=False)
fig_seg.add_trace(go.Bar(x=segment_summary.index, y=segment_summary['total_profit'], name="Profit", marker_color=colors[1]), secondary_y=False)

fig_seg.add_trace(go.Scatter(x=segment_summary.index, y=segment_summary['cumperc_sales'], name="% Cumulative Sales", line=dict(color='hotpink', width=2), mode='lines+markers'), secondary_y=True)
fig_seg.add_trace(go.Scatter(x=segment_summary.index, y=segment_summary['cumperc_profit'], name="% Cumulative Profit", line=dict(color='yellow', width=2), mode='lines+markers'), secondary_y=True)

fig_seg.update_layout(
    title_text="Total Sales and Profit by Customer Segment",
    title_x=0.5,
    barmode='group',
    font=dict(size=14),
    template=theme,
    xaxis_title="Segment",
    legend=dict(orientation="h", y=-0.2, x=0.25),
    height=600,
    width=676
)
fig_seg.update_yaxes(title_text="Sales / Profit", secondary_y=False)
fig_seg.update_yaxes(title_text="Cumulative (%)", secondary_y=True, range=[0, 110])
fig_seg.show()

### Distribution
- **Consumer**: 1.16M sales, 134K profit.
- **Corporate**: 706K sales, 92K profit.
- **Home Office**: 430K sales, 60K profit.

### Cumulative
- Consumer + Corporate: ~80% revenue/profit (Pareto). Prioritize these.

## Product Performance Analysis

Top 5 by Sales

In [29]:
top_products = df.groupby('Product Name').agg(
    total_sales=('Sales', 'sum'),
    total_profit=('Profit', 'sum')
).sort_values(by='total_sales', ascending=False).head(5).sort_values(by='total_sales', ascending=True)

Plot

In [30]:
fig_prod = go.Figure()

fig_prod.add_trace(go.Bar(y=top_products.index, x=top_products['total_sales'], name='Revenue', orientation='h', marker_color='#5dade2', text=[f"{v:,.0f}" for v in top_products['total_sales']], textposition='outside'))
fig_prod.add_trace(go.Bar(y=top_products.index, x=top_products['total_profit'], name='Profit', orientation='h', marker_color='#58d68d', text=[f"{v:,.0f}" for v in top_products['total_profit']], textposition='outside'))

fig_prod.update_layout(
    title='TOP 5 PRODUCTS WITH HIGHEST REVENUE & PROFIT',
    xaxis_title='Value ($)',
    yaxis_title='Product Name',
    barmode='group',
    bargap=0.25,
    template=theme,
    font=dict(size=14),
    height=550,
    legend=dict(orientation='h', y=-0.2, x=0.25)
)
xmin = min(top_products['total_profit'].min(), 0)
xmax = top_products['total_sales'].max() * 1.15
fig_prod.update_xaxes(range=[xmin, xmax])
fig_prod.show()

- **Canon Copier**: 61.6K sales, 25.2K profit.
- **Cisco TelePresence**: High sales, but loss (pricing/costs?).

## Correlation Analysis

Correlation matrix

In [31]:
corr_matrix = df[['Sales', 'Profit', 'Discount']].corr().round(2)

Heatmap

In [32]:
fig_corr = px.imshow(
    corr_matrix,
    text_auto=True,
    color_continuous_scale='RdBu_r',
    title='CORRELATION HEATMAP BETWEEN SALES, PROFIT AND DISCOUNT',
    aspect='auto',
    template=theme
)
fig_corr.update_layout(
    coloraxis_colorbar=dict(title="Correlation"),
    font=dict(size=14),
    title_x=0.5,
    height=600,
    margin=dict(l=50, r=50, t=80, b=50)
)
fig_corr.show()

### Correlations
- Sales-Profit: 0.48 (positive; sales drive profit).
- Discount-Profit: -0.22 (negative; high discounts erode profit).
- Sales-Discount: -0.03 (negligible).
> Drive sales, but manage discounts carefully.